### PR TITLE
Add SDPO (Self-Distillation Policy Optimization) trainer

### LINKED ISSUE #4929: SDPO: Reinforcement Learning via Self-Distillation
### Feature request

In the last 5 days I see a lot of different apporaches on self-distillation where one model is both the teacher and student. Especially interesting is this paper: https://arxiv.org/abs/2601.20802

Is there any way of we can implement a trainer like this in trl ? I assume this would be part of the experimental trainers knowing the refactoring stuff you did recently in trl.

### Motivation

Would love to experiment with these approaches, especially on smaller models (tiny models)

### Your contribution

I can try to implement SDPO if I get any tips of how you imagine this being implemented.

### PR DESCRIPTION
**Pull request by MengAiDev**
Implements SDPO algorithm from arxiv.org/abs/2601.20802. SDPO augments on-policy optimization with self-distillation from the model's own high-reward trajectories, converting tokenized feedback into a dense learning signal.

- Add SDPOConfig with distillation parameters (alpha, topk, ema_update_rate, etc.)
- Add SDPOTrainer extending GRPOTrainer with self-distillation loss
- Add comprehensive tests for SDPOConfig and SDPOTrainer
- Add example script demonstrating SDPO usage

Fixes #4929

---

> [!NOTE]
> **Medium Risk**
> Introduces new experimental training algorithms (online rollout + self-distillation, EMA teacher syncing, reward-driven reprompting) that can affect training correctness and distributed behavior, though changes are mostly additive and isolated under `trl.experimental`.
> 
> **Overview**
> Adds a new **experimental self-distillation stack** (`SelfDistillationConfig`, `SelfDistillationMixin`, `OnlineRolloutMixin`, `BaseSelfDistillationTrainer`) to support rollout reuse, reward scoring/normalization, self-distillation losses (token/logit-level + optional IS clipping), and callback hooks/diagnostics.
> 
> Introduces two new trainers: **`SDPOTrainer`/`SDPOConfig`** implementing online SDPO with successful-rollout/feedback-based teacher reprompting and optional EMA teacher synchronization, and **`SDFTTrainer`/`SDFTConfig`** for on-policy self-distilled fine-tuning using teacher-conditioned prompts and optional PEFT adapter EMA teacher.
> 
> Adds example scripts (`trl/experimental/sdpo/sdpo.py`, `trl/experimental/sdft/sdft.py`), new docs pages and paper index entries, and extensive tests covering training flows, callback payloads, PEFT EMA behavior, masking/attention correctness, and diagnostic warnings.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bf4cc67ac32ab9418b8239665d15458ab77c00cc. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
@MengAiDev I have cleaned up the structure and docs and tests. Next we need to address the main TODOs regarding the teacher logits.

**Comment by kashif** [comment]
cc @jonhue here is a port of SDPO for TRL

**Comment by jonhue** [comment]
@MengAiDev @kashif Thanks so much for implementing this!! Let's coordinate with @Shekswess and #4941. It might be cleanest to have one implementation for SDFT & SDPO ("self-distillation") since both are algorithmically the same and they differ only in whether data is offline or online.

**Comment by kashif** [comment]
agree! lets try that if its ok for you @MengAiDev

**Comment by Shekswess** [comment]
Wohoo !
This is really awesome, bravo legends @kashif @jonhue @MengAiDev. Maybe we should also then have the offline version of the trainer, knowing that some folks (like me that are GPU poor hahahahaha) can experiment with the approaches

**Comment by LeonEricsson** [comment]
Regarding the discussion on how to combine SDFT/SDPO PRs:

This PR inherits from GRPOTrainer, while the SDFT PR modifies it in place. Both approaches carry baggage from GRPOTrainer that isn’t necessarily applicable to SDPO/SDFT — but this also provides a nice playground for experimentation.

The tradeoff with inheritance is less control, but I like how it nicely isolates SDPO’s key contributions and exposes relevant hparams clearly. If future research demands more flexibility,  we can revisit and consider breaking out SDPO into its own trainer. 

If we proceed with this PR’s approach, extending it to cover the offline case should, at first glance, just require modifying the _build_teacher_inputs function.​​​​​​​​​​​​​​​​

**Comment by qgallouedec** [comment]
That a good point Leon, I need to review the PR carefully, but in general, I’d rather isolate first and abstract later, if needed. (abstractions are easy to do, hard to undo)

**Comment by Shekswess** [comment]
@qgallouedec @LeonEricsson if you see my implementation https://github.com/huggingface/trl/pull/4941#issuecomment-3833298159, of the offline SDFT I think it can be really really improved, tried to follow the official code from the authors with small modifications, feel free to ping us on how we can make these stuff better. Cannot wait to start to experiment hehehehe

**Comment by niksdagr8** [comment]
Any progress on this is much appreciated

File: docs/source/sdft_trainer.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
+# SDFT
+
+Self-Distilled Fine-Tuning (SDFT) is described in [Self-Training with On-Policy Self-Distillation for Language Model Alignment](https://huggingface.co/papers/2601.19897).
+
+The TRL implementation adapts SDFT to the experimental trainer API while reusing the shared self-distillation infrastructure also used by SDPO.
+
+In the current TRL implementation:
+
+- SDFT uses an explicit `ref_model` teacher
```
Comment (qgallouedec) [inline docs/source/sdft_trainer.md]: is ref_model the teacher? or are these two separate models?

File: docs/source/sdft_trainer.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    model="Qwen/Qwen2.5-1.5B-Instruct",
+    ref_model="Qwen/Qwen2.5-1.5B-Instruct",
+    args=training_args,
+    train_dataset=dataset,
+)
+trainer.train()
+```
+
+To generate from the teacher-conditioned prompt instead of the student prompt, set `generate_from_teacher=True`.
+To customize how the teacher prompt is built, set `teacher_prompt_template` on `SDFTConfig`.
```
Comment (qgallouedec) [inline docs/source/sdft_trainer.md]:
```suggestion
To customize how the teacher prompt is built, set `teacher_prompt_template` on [`SDFTConfig`].
```

File: docs/source/sdft_trainer.md (hunk 1 above)
Comment (kashif) [reply docs/source/sdft_trainer.md]: yes its the teacher

File: trl/experimental/self_distillation/self_distillation_mixin.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            if self.args.distillation_add_tail:
+                topk_student_log_probs = self._add_tail(topk_student_log_probs)
+                topk_teacher_log_probs = self._add_tail(topk_teacher_log_probs)
+            else:
+                topk_student_log_probs = self._renorm_topk_log_probs(topk_student_log_probs)
+                topk_teacher_log_probs = self._renorm_topk_log_probs(topk_teacher_log_probs)
+
+            per_token_loss = self._compute_divergence(
+                topk_student_log_probs, topk_teacher_log_probs, self.args.distillation_alpha
+            )
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/experimental/self_distillation/self_distillation_mixin.py:163]: not a bug

File: docs/source/paper_index.md [hunk 4]
```diff
@@ -1630,6 +1630,87 @@ trainer.train()
...
+- `prompt`
+- `privileged_context` for optional environment feedback
+
+For more details, see the [SDPO Trainer documentation](sdpo_trainer).
+
+### Self-Training with On-Policy Self-Distillation for Language Model Alignment
+
+**📜 Paper**: https://huggingface.co/papers/2601.19897
+
+Self-Distilled Fine-Tuning (SDFT) performs on-policy self-distillation by generating completions during training, then distilling an explicit teacher-conditioned view of those same completions back into the student. In TRL, SDFT uses a shared self-distillation core with SDPO while keeping its own explicit `ref_model` teacher and dataset-provided privileged context.
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: > its own explicit `ref_model` teacher

for more clarity, and to avoid confusion with ref model in DPO/GRPO/RLOO etc, I'd use `teacher_model`

File: docs/source/sdft_trainer.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+```python
+from datasets import Dataset
+
+from trl.experimental.sdft import SDFTConfig, SDFTTrainer
+
+dataset = Dataset.from_dict(
+    {
+        "prompt": ["Solve 2+2."],
+        "privileged_context": ["Example answer: 4."],
```
Comment (qgallouedec) [inline docs/source/sdft_trainer.md]: shouldn't we use conversational format?

File: docs/source/sdpo_trainer.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# SDPO
+
+Self-Distillation Policy Optimization (SDPO) was introduced in [Reinforcement Learning via Self-Distillation](https://huggingface.co/papers/2601.20802) by Jonas Hübotter, Frederike Lübeck, Lejs Behric, Anton Baumann, Marco Bagatella, Daniel Marta, Ido Hakimi, Idan Shenfeld, Thomas Kleine Buening, Carlos Guestrin, and Andreas Krause.
```
Comment (qgallouedec) [inline docs/source/sdpo_trainer.md]:
```suggestion
Self-Distillation Policy Optimization (SDPO) was introduced in [Reinforcement Learning via Self-Distillation](https://huggingface.co/papers/2601.20802) by [Jonas Hübotter](https://huggingface.co/jonhue), Frederike Lübeck, Lejs Behric, [Anton Baumann](https://huggingface.co/antonbaumann), Marco Bagatella, Daniel Marta, Ido Hakimi, Idan Shenfeld, Thomas Kleine Buening, Carlos Guestrin, and Andreas Krause.
```

File: trl/trainer/callbacks.py [hunk 7]
```diff
@@ -756,3 +756,122 @@ def on_train_end(self, args: TrainingArguments, state: TrainerState, control: Tr
             save_directory = f"{args.output_dir}/bema"
             self.running_model.save_pretrained(save_directory)
             logger.info(f"Saved BEMA model to {save_directory}")
+
+
+class PEFTAdapterEMACallback(TrainerCallback):
```
Comment (qgallouedec) [inline trl/trainer/callbacks.py]: I recommend moving this callback to `trl.experimental.peft_adapter_ema_callback.py` instead since it's only compatible with distillation trainers, which we don't have in trl "stable" yet

File: trl/experimental/self_distillation/base_self_distillation_trainer.py [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+logger = get_logger(__name__)
+
+
+class BaseSelfDistillationTrainer(OnlineRolloutMixin, SelfDistillationMixin, _BaseTrainer):
+    """Shared scaffold for experimental self-distillation trainers without GRPO inheritance."""
+
+    config_cls = SelfDistillationConfig
+    _tag_names = ["trl", "self-distillation"]
+    _name = "SelfDistillation"
```
Comment (qgallouedec) [inline trl/experimental/self_distillation/base_self_distillation_trainer.py]: If I'm not mistaken, these will be overridden by the 2 childs, so I think you can remove them

File: trl/experimental/sdpo/sdpo_trainer.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    SDPO augments on-policy optimization with self-distillation from the model's own high-reward trajectories. It
+    converts tokenized feedback into a dense learning signal without any external teacher or explicit reward model.
+    SDPO treats the current model conditioned on feedback as a self-teacher and distills its feedback-informed
+    next-token predictions back into the policy.
+    """
+
+    config_cls = SDPOConfig
+    _tag_names = ["trl", "sdpo"]
+    _name = "SDPO"
+
```
Comment (qgallouedec) [inline trl/experimental/sdpo/sdpo_trainer.py:262]: can you add `_paper` as well?

eg for bco

```python
_paper = {
        "title": "Binary Classifier Optimization for Large Language Model Alignment",
        "id": "2404.04656",
        # docstyle-ignore
        "citation": textwrap.dedent("""\
            @article{jung2024binary,
                title        = {{Binary Classifier Optimization for Large Language Model Alignment}},
                author       = {Seungjae Jung and Gunsoo Han and Daniel Wontae Nam and Kyoung{-}Woon On},
                year         = 2024,
                eprint       = {arXiv:2404.04656}
            }"""),
    }
```

Comment (qgallouedec) [reply trl/experimental/sdpo/sdpo_trainer.py:262]: and same for sdft

**Comment by qgallouedec** [comment]
@codex review

File: trl/experimental/self_distillation/teacher_context.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+from ...trainer.utils import pad
+
+
+def extract_last_user_text(prompt: list[dict[str, Any]]) -> str:
+    """Extract the text content from the last message in a conversational prompt."""
+    last_message = prompt[-1]
+    content = last_message.get("content", "")
+    if isinstance(content, list):
+        return " ".join(part.get("text", "") for part in content if part.get("type") == "text")
+    return content
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/self_distillation/teacher_context.py:39]: I don't think that's a very important feature right now

**Comment by qgallouedec** [comment]
Just running a final @codex review to see if there is any critical bug, otherwise we're good to merge

File: trl/experimental/self_distillation/self_distillation_config.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+    This class contains only the arguments that are specific to the shared self-distillation stack. For the full set of
+    generic training arguments, refer to [`~transformers.TrainingArguments`] via
+    [`trl.trainer.base_config._BaseConfig`].
+
+    Parameters:
+        > Parameters that control generation and rollout reuse
+
+        model_init_kwargs (`dict[str, Any]`, *optional*):
+            Keyword arguments used when the `model` argument is passed as a string.
+        max_prompt_length (`int` or `None`, *optional*, defaults to `512`):
```
Comment (qgallouedec) [inline trl/experimental/self_distillation/self_distillation_config.py:37]: not necessarily for this PR, but I think that in general, truncating the prompt isn't a good idea, because it truncates the generation prompt (`<|im_end|><|im_start|>assistant\n`), so the generation basically completes the user query, instead of answering it, which does make sense, and I don't think there is anything to learn from it. Unless there is a good reason to keep it, I recommend removing it

**Comment by Neelectric** [comment]
It looks like `SDFTConfig` has `True` as the [default parameter](https://github.com/huggingface/trl/blob/bad33f28fafedb71d8e4c745c2d70707e9bac7a4/trl/experimental/sdft/sdft_config.py#L39) for `disable_dropout`, while `SelfDistillationConfig` uses `False` as the [default parameter](https://github.com/huggingface/trl/blob/bad33f28fafedb71d8e4c745c2d70707e9bac7a4/trl/experimental/self_distillation/self_distillation_config.py#L83).

The latter more closely matches the reference implementation by Shenfeld et al., where `disable_dropout` also uses `False` as the [default parameter](https://github.com/idanshen/Self-Distillation/blob/aa70efaf306c6d4a0fbeed9975d6fee967fbdef9/distil_config.py#L40), and I did not find any overwrites for this elsewhere in their repo. Should `SDFTConfig` also default to `False` for consistency?

**Comment by 1dividedby0** [comment]
Does this implementation do online feedback? I see that the privileged context is generated offline as part of the dataset. Are there any plans to make this online (i.e. using the privileged context that is generated after the rollout)?
